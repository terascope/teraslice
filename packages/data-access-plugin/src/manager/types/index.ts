import DataType from './data-type';
import Mutation from './mutation';
import Query from './query';
import Role from './role';
import Space from './space';
import User from './user';
import View from './view';

export default [
    'scalar JSON',
    'scalar DateTime',
    DataType,
    Role,
    Space,
    User,
    View,
    Query,
    Mutation,
];
