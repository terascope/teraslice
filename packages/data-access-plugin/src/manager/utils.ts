import { Request } from 'express';
import * as ts from '@terascope/utils';
import * as apollo from 'apollo-server-express';
import { User, ACLManager, ModelName, AnyModel } from '@terascope/data-access';

export function forEachModel(fn: (model: ModelName) => void) {
    const models: ModelName[] = ['User', 'Role', 'DataType', 'Space', 'View'];
    return models.forEach(model => fn(model));
}

export function formatError(err: any) {
    if (err && err.extensions != null) return err;

    const { statusCode, message, code } = ts.parseErrorInfo(err);

    let error: any;

    if (statusCode >= 400 && statusCode < 500) {
        if (statusCode === 422) {
            error = new apollo.ValidationError(message);
        } else if (statusCode === 401) {
            error = new apollo.AuthenticationError(message);
        } else if (statusCode === 403) {
            error = new apollo.ForbiddenError(message);
        } else {
            error = new apollo.UserInputError(message);
        }

        if (err && err.stack) error.stack = err.stack;
    } else {
        error = apollo.toApolloError(err, code);
    }

    return error;
}

export function setLoggedInUser(req: Request, user: User, storeInSession = true): void {
    if (storeInSession) {
        ts.set(req, 'session.v2User', user);
    }
    ts.set(req, 'v2User', user);
}

export function logoutUser(req: Request): void {
    // @ts-ignore
    if (req.session) delete req.session.v2User;
    // @ts-ignore
    if (req.v2user) delete req.v2user;
    // @ts-ignore
    if (ts.isFunction(req.logout)) req.logout();
}

export function getLoggedInUser(req: Request): User | null {
    const user = ts.get(req, 'v2User', ts.get(req, 'session.v2User'));
    // the user must be the latest type
    if (user && user.type && user.id) {
        return user;
    }
    return null;
}

interface Creds {
    token?: string;
    username?: string;
    password?: string;
}

function isSameUser(user: User, creds: Creds): boolean {
    // if creds then compare tokens
    if (!ts.isEmpty(creds)) {
        if (creds.token && user.api_token === creds.token) return true;
        return false;
    }
    // if just user return true
    return true;
}

export async function login(manager: ACLManager, req: Request, storeInSession = true): Promise<User> {
    const loggedInUser = getLoggedInUser(req);
    const creds = getCredentialsFromReq(req);

    if (loggedInUser && isSameUser(loggedInUser, creds)) {
        // fetch and update the user again
        return manager.findUser({ id: loggedInUser.id }, loggedInUser);
    }

    const user = await manager.authenticate(creds);
    setLoggedInUser(req, user, storeInSession);
    return user;
}

export function getCredentialsFromReq(req: Request): Creds {
    const queryToken: string = ts.get(req, 'query.token');
    if (queryToken) return { token: queryToken };

    const authToken: string = ts.get(req, 'headers.authorization');
    if (!authToken) return {};

    let [part1, part2] = authToken.split(' ');
    part1 = ts.trim(part1);
    part2 = ts.trim(part2);

    if (part1 === 'Token') {
        return { token: ts.trim(part2) };
    }

    if (part1 === 'Basic') {
        const parsed = Buffer.from(part2, 'base64').toString('utf8');
        const [username, password] = parsed.split(':');
        return { username, password };
    }

    return {};
}

export async function findAll<T extends AnyModel>(ids: string[] | undefined, fn: (query: string) => Promise<T[]>): Promise<T[]> {
    const _ids = ts.castArray(ids || []).filter(id => !!id);
    if (!_ids.length) return [];

    const query = `id: (${_ids.join(' OR ')})`;
    return fn(query);
}
