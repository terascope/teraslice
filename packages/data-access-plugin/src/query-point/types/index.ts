import User from './user';
import Query from './query';
import Geo from './geo-type';

// TODO: query and User should probably not be here
export default ['scalar DateTime', 'scalar JSON', Query, User, Geo];
