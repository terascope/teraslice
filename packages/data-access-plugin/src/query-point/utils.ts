
import * as apollo from 'apollo-server-express';
import { SelectionSetNode, ArgumentNode } from 'graphql';
import { ParsedJoinFields } from './interfaces';

interface MyDict {
    [key: string]: boolean;
}

function getJoinKeys(ast: ArgumentNode[]) {
    for (const arg of ast) {
        if (arg.name.value === 'join') {
            if (arg.value.kind === 'ListValue') {
                return arg.value.values.map((obj) => {
                    if (obj.kind === 'StringValue') {
                        return obj.value;
                    }
                    throw new apollo.UserInputError('values inside a join parameter must be a string');
                });
            }
            throw new apollo.UserInputError('join parameters must be an array of strings');
        }
    }
    return [];
}

function checkQuery(ast: SelectionSetNode) {
    const currentFields: MyDict = {};
    const joinKeys: string[] = [];

    for (const selector of ast.selections) {
        if (selector.kind === 'Field') {
            const { name: { value: field } } = selector;
            currentFields[field] = true;
            if (selector.arguments) {
                // @ts-ignore
                joinKeys.push(...getJoinKeys(selector.arguments));
            }
            if (selector.selectionSet) checkQuery(selector.selectionSet);
        }
    }

    for (const originalKey of joinKeys) {
        const { origin: key } = parseJoinField(originalKey);
        if (!currentFields[key]) {
            const errMsg = `Invalid join on field "${key}", you may only join on a field that is being returned in a query`;
            throw new apollo.UserInputError(errMsg);
        }
    }
}

export function parseJoinField(fieldParams: string): ParsedJoinFields {
    // any chars between a | and a whitespace or :
    let field = fieldParams;
    let extraParams: undefined | string;
    const regex = /\|([^: ]+)/;
    const hasExtraParams = field.match(regex);

    if (hasExtraParams) {
        [, extraParams] = hasExtraParams;
        field = field.replace(`|${extraParams}`, '');
    }

    const selectorTargets = field.split(':');
    const origin = selectorTargets[0];
    const target = selectorTargets[1] || origin;

    return { origin, target, extraParams };
}


export function validateJoins(query: string) {
    const { definitions } = apollo.gql(query);
    definitions.forEach((type) => {
        if (type.kind === 'OperationDefinition') {
            checkQuery(type.selectionSet);
        }
    });
}
