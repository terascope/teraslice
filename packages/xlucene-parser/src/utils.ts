import {
    isString,
    isEmpty,
    isPlainObject,
    TSError,
    isWildCardString,
    debugLogger,
    primitiveToString,
    isPrimitiveValue,
    toBoolean,
    toNumber,
    toIntegerOrThrow,
    toFloatOrThrow,
    getTypeOf,
    isBooleanLike,
    isNotNil,
    isIP,
    isCIDR,
    isIPv6,
    getFirstIPInCIDR,
    getLastIPInCIDR,
    toCIDR,
    isValidDate,
    getValidDate
} from '@terascope/utils';

import {
    xLuceneFieldType, xLuceneVariables, CoordinateTuple,
    Maybe
} from '@terascope/types';
import * as i from './interfaces.js';

export const logger = debugLogger('xlucene-parser');

function _getType(node: unknown): i.NodeType | undefined {
    if (!node || typeof node !== 'object') return;
    return (node as any).type || undefined;
}

export function isLogicalGroup(node: unknown): node is i.LogicalGroup {
    return _getType(node) === i.NodeType.LogicalGroup;
}

export function isConjunction(node: unknown): node is i.Conjunction {
    return _getType(node) === i.NodeType.Conjunction;
}

export function isNegation(node: unknown): node is i.Negation {
    return _getType(node) === i.NodeType.Negation;
}

export function isFieldGroup(node: unknown): node is i.FieldGroup {
    return _getType(node) === i.NodeType.FieldGroup;
}

export function isExists(node: unknown): node is i.Exists {
    return _getType(node) === i.NodeType.Exists;
}

export function isRange(node: unknown): node is i.Range {
    return _getType(node) === i.NodeType.Range;
}

export function isFunctionNode(node: unknown): node is i.FunctionNode {
    return _getType(node) === i.NodeType.Function;
}

export function isRegexp(node: unknown): node is i.Regexp {
    return _getType(node) === i.NodeType.Regexp;
}

export function isWildcard(node: unknown): node is i.Wildcard {
    return _getType(node) === i.NodeType.Wildcard;
}

export function isWildcardField(node: unknown): boolean {
    return !!(node && isWildCardString((node as any).field));
}

export function isTerm(node: unknown): node is i.Term {
    return _getType(node) === i.NodeType.Term;
}

export function isTermList(node: unknown): node is i.TermList {
    return _getType(node) === i.NodeType.TermList;
}

export function isEmptyNode(node: unknown): node is i.EmptyNode {
    return isEmpty(node) || (_getType(node) === i.NodeType.Empty);
}

export function isStringDataType(node: unknown): node is i.StringDataType {
    return !!(node && (node as any).field_type === 'string');
}

export const numberDataTypes: xLuceneFieldType[] = [
    xLuceneFieldType.Integer, xLuceneFieldType.Float
];

export function isNumberDataType(node: unknown): node is i.NumberDataType {
    return !!(node && numberDataTypes.includes((node as any).field_type));
}

export function isBooleanDataType(node: unknown): node is i.BooleanDataType {
    return !!(node && (node as any).field_type === 'boolean');
}

export function getField(node: unknown): string | undefined {
    if (!node) return;
    if (!(node as any).field || !isString((node as any).field)) return;
    return (node as any).field;
}

/** term level queries with field (string|null)  */
export const termTypes: readonly i.NodeType[] = [
    i.NodeType.Term,
    i.NodeType.Regexp,
    i.NodeType.Range,
    i.NodeType.Wildcard,
    i.NodeType.Function,
    i.NodeType.TermList,
];

export function isTermType(node: unknown): node is i.TermLikeNode {
    return !!(node && termTypes.includes((node as any).type));
}

/** logical group or field group with flow */
export const groupTypes: i.NodeType[] = [i.NodeType.LogicalGroup, i.NodeType.FieldGroup];

export function isGroupLike(node: unknown): node is i.GroupLikeNode {
    return !!(node && groupTypes.includes((node as any).type));
}

/**
 * Validate and normalize a variables object.
 * 
 * @param obj - The variables object to validate
 * @returns A normalized copy of the variables object
 * @throws {TSError} If the variables object is invalid
 */
export function validateVariables(obj: xLuceneVariables): xLuceneVariables {
    if (!isPlainObject(obj)) {
        throw new TSError('Invalid xLuceneVariables configuration provided, it must be an object', {
            context: { safe: true },
            statusCode: 400
        });
    }
    return { ...obj };
}

/**
 * Resolve a field value or array of field values using the provided variables.
 * 
 * This function handles both literal values and variable references, resolving
 * variables to their actual values from the variables object.
 * 
 * @param value - The field value(s) to resolve
 * @param variables - Object containing variable values
 * @param allowNil - Whether to allow undefined/null values
 * @returns The resolved value(s)
 */
export function getFieldValue<T>(
    value: i.FieldValue<T>,
    variables: xLuceneVariables,
    allowNil: true
): Maybe<T>;
export function getFieldValue<T>(
    value: i.FieldValue<T>,
    variables: xLuceneVariables,
    allowNil?: boolean
): T;
export function getFieldValue<T>(
    value: i.FieldValue<T>[],
    variables: xLuceneVariables,
    allowNil?: boolean
): T[];
export function getFieldValue<T>(
    value: i.FieldValue<T> | i.FieldValue<T>[],
    variables: xLuceneVariables,
    allowNil?: boolean
): (T | Maybe<T>)|(T[]) {
    if (Array.isArray(value)) {
        return value
            .map((val) => getFieldValue<any>(val, variables, true))
            .filter(isNotNil);
    }

    if (value.type === 'variable') {
        const varValue = variables[value.value];

        /* a variable returning undefined is different from returning null
        * undefined will signal dataframe and document matcher to
        * return false for this logic operator node, this is to allow
        * AND and OR operations to correctly return expected results
        * if joining or partial data. Returning null would return incorrect
        * results as it is to permissive in an OR statement
        */
        if (varValue === undefined) {
            return undefined;
        }

        if (varValue != null) {
            return variables[value.value] as T;
        }

        if (allowNil) return null;
    }

    if (value.value != null) {
        return value.value as T;
    }

    if (allowNil) return null;

    return undefined;
}

export function isInfiniteValue(input?: number | string): boolean {
    return input === '*' || input === Number.NEGATIVE_INFINITY || input === Number.POSITIVE_INFINITY;
}

export function isInfiniteMin(min?: number | string): boolean {
    if (min == null) return false;
    return min === '*' || min === Number.NEGATIVE_INFINITY;
}

export function isInfiniteMax(max?: number | string): boolean {
    if (max == null) return false;
    return max === '*' || max === Number.POSITIVE_INFINITY;
}

export interface ParsedRange {
    gte?: number | string;
    gt?: number | string;
    lte?: number | string;
    lt?: number | string;
}

/**
 * Parse a range node into a simple object with comparison operators.
 * 
 * @param node - The range node to parse
 * @param variables - Variables for resolving variable references
 * @param excludeInfinite - Whether to exclude infinite values (* or Â±Infinity)
 * @returns Object with comparison operators and their values
 */
export function parseRange(
    node: i.Range, variables: xLuceneVariables, excludeInfinite = false
): ParsedRange {
    const results: Partial<ParsedRange> = {};

    const leftValue = getFieldValue(node.left.value, variables);
    if (!excludeInfinite || !isInfiniteValue(leftValue)) {
        results[node.left.operator] = leftValue;
    }

    if (node.right) {
        const rightValue = getFieldValue(node.right.value, variables);
        if (!excludeInfinite || !isInfiniteValue(rightValue)) {
            results[node.right.operator] = rightValue;
        }
    }
    return results;
}

export function coordinateToXlucene(cord: CoordinateTuple): string {
    // tuple is [lon, lat], need to return "lat, lon"
    return `"${cord[1]}, ${cord[0]}"`;
}

type CoerceValueFns = Partial<Record<xLuceneFieldType, ((value: any) => any)>>;
export const coerceValueFns: CoerceValueFns = Object.freeze({
    [xLuceneFieldType.AnalyzedString]: primitiveToString,
    [xLuceneFieldType.String]: primitiveToString,
    [xLuceneFieldType.Boolean](value) {
        if (!isBooleanLike(value)) {
            throw new Error(`Expected ${value} (${getTypeOf(value)}) to be in a boolean like format`);
        }
        return toBoolean(value);
    },
    [xLuceneFieldType.Date](value) {
        if (!isPrimitiveValue(value)) {
            throw new Error(`Expected ${value} (${getTypeOf(value)}) to be in a date like format`);
        }

        if (isValidDate(value)) return value;

        const coerced = getValidDate(value, new Date());
        if (coerced) return coerced;

        return value;
    },
    [xLuceneFieldType.Number](value) {
        const numValue = toNumber(value);
        if (Number.isNaN(numValue)) {
            throw new Error(`Expected ${value} (${getTypeOf(value)}) to be in a number like format`);
        }
        return numValue;
    },
    [xLuceneFieldType.Float]: toFloatOrThrow,
    [xLuceneFieldType.Integer]: toIntegerOrThrow,
    [xLuceneFieldType.IP](value) {
        if (typeof value !== 'string') {
            throw new Error(`Expected ${value} (${getTypeOf(value)}) to be in a IP like format`);
        }
        return value;
    },
});

/**
 * Create a coercion function for the specified field type.
 * 
 * @param fieldType - The xLucene field type
 * @returns A function that coerces values to the specified type
 */
export function makeCoerceFn(fieldType: xLuceneFieldType | undefined): (v: any) => any {
    if (!fieldType || !(fieldType in coerceValueFns)) return (v) => v;
    const coerceFn = coerceValueFns[fieldType]!;
    return coerceFn;
}

/**
 * Convert an IP address or CIDR block term into a range query.
 * 
 * @param node - The term node containing the IP value
 * @param value - The IP address or CIDR block string
 * @returns A range node representing the IP range
 */
export function createIPRangeFromTerm(node: i.Term, value: string): i.Range {
    const { start, end } = parseIPRange(value);

    return {
        type: i.NodeType.Range,
        field: node.field,
        field_type: node.field_type,
        left: {
            operator: 'gte',
            field_type: xLuceneFieldType.IP,
            value: {
                type: 'value',
                value: start,
            }
        },
        right: {
            operator: 'lte',
            field_type: xLuceneFieldType.IP,
            value: {
                type: 'value',
                value: end,
            },
        }
    };
}

function parseIPRange(val: string): { start: string; end: string } {
    const cidrBlock = makeCidr(val);

    return {
        start: getFirstIPInCIDR(cidrBlock),
        end: getLastIPInCIDR(cidrBlock)
    };
}

function makeCidr(val: string): string {
    if (isCIDR(val)) return val;

    if (isIP(val)) {
        // CIDR notation for a single ip
        if (isIPv6(val)) return toCIDR(val, 128);
        return toCIDR(val, 32);
    }

    throw new Error(`Invalid value ${val}, could not convert to ip_range`);
}
