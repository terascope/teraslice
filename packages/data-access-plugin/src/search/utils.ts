import * as ts from '@terascope/utils';
import { Client } from 'elasticsearch';
import * as da from '@terascope/data-access';
import * as i from './interfaces';

/**
 * Search elasticsearch in a teraserver backwards compatible way
 */
export function makeSearchFn(
    client: Client,
    accessConfig: da.DataAccessConfig,
    logger: ts.Logger
): i.SearchFn {
    const searchAccess = new da.SearchAccess(accessConfig, logger);

    return async (query: da.InputQuery) => searchAccess.performSearch(client, query);
}
