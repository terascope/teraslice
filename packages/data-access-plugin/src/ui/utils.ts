import * as ts from '@terascope/utils';

export function getErrorInfo(err: any): { message: string, statusCode: number } {
    // apollo likes to hide the original error
    const hiddenErr = ts.get(err, 'networkError.result.errors[0]');
    if (!hiddenErr) {
        const { message, statusCode } = ts.parseErrorInfo(err);
        return { message, statusCode };
    }
    const message = ts.get(hiddenErr, 'message');
    const code = ts.get(hiddenErr, 'extensions.code');
    if (code === 'UNAUTHENTICATED') {
        return {
            statusCode: 401,
            message,
        };
    }

    for (const [statusCode, errorCode] of Object.entries(ts.STATUS_ERROR_CODES)) {
        if (code === errorCode) {
            return {
                statusCode: ts.toNumber(statusCode),
                message,
            };
        }
    }
    return {
        message,
        statusCode: ts.get(err, 'networkError.result.statusCode', 500)
    };
}
