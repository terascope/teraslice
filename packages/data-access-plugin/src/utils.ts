import { Client } from 'elasticsearch';
import { Request, Response } from 'express';
import { Context } from '@terascope/job-components';
import * as ts from '@terascope/utils';

export type ErrorHandlerFn = (req: Request, res: Response, fn: (...args: any[]) => Promise<any>|any) => Promise<void>;

export function makeErrorHandler(reason: string, logger: ts.Logger, requireSafe = false): ErrorHandlerFn {
    return async (req, res, fn) => {
        try {
            await fn();
        } catch (error) {
            const statusCode = ts.getErrorStatusCode(error);

            logger.error(error, reason, {
                path: req.originalUrl,
                query: req.query
            });

            const resp: any = {
                error: ts.stripErrorMessage(error, reason, requireSafe)
            };

            const user = ts.get(req, 'v2User', { type: 'USER' });
            if (user.type !== 'USER') {
                resp.debug = {
                    timestamp: ts.makeISODate()
                };

                if (ts.isString(error)) {
                    resp.debug.message = error;
                } else {
                    resp.debug.message = error.message;
                    resp.debug.stack = error.stack;
                    resp.debug.statusCode = statusCode;
                }
            }

            res.status(statusCode).json(resp);
        }
    };
}

export function getESClient(context: Context, connection: string): Client {
    return context.foundation.getConnection({
        type: 'elasticsearch',
        endpoint: connection,
        cached: true
    }).client;
}
