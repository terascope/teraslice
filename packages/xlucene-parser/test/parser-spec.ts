import 'jest-extended';
import { TSError, times, toString } from '@terascope/utils';
import { xLuceneFieldType } from '@terascope/types';
import allTestCases, { filterNilTestCases } from './cases';
import {
    Parser, NodeType, FieldValue, TermLikeNode
} from '../src';

describe('Parser', () => {
    for (const [key, testCases] of Object.entries(allTestCases)) {
        describe(`when testing ${key.replace('_', ' ')} queries`, () => {
            describe.each(testCases)('given query %s', (query, msg, ast, typeConfig, variables) => {
                if (variables) {
                    it(`should be able to parse ${msg} with variables ${toString(variables)}`, () => {
                        const parser = new Parser(query, {
                            type_config: typeConfig,
                        }).resolveVariables(variables);

                        expect(parser.ast).toMatchObject(ast);
                    });
                } else {
                    it(`should be able to parse ${msg}`, () => {
                        const parser = new Parser(query, {
                            type_config: typeConfig,
                        });
                        expect(parser.ast).toMatchObject(ast);
                    });
                }
            });
        });
    }

    describe('when parser has filterNilVariables set true', () => {
        for (const [key, testCases] of Object.entries(filterNilTestCases)) {
            describe(`when testing ${key.replace('_', ' ')} queries`, () => {
                describe.each(testCases)('given query %s', (query, msg, ast, typeConfig, variables, resolved) => {
                    it(`should be able to parse ${msg} ${variables ? `with variables ${toString(variables)}` : ''}`, () => {
                        const parser = new Parser(query, {
                            type_config: typeConfig,
                            filterNilVariables: true,
                            variables
                        });

                        expect(parser.ast).toMatchObject(ast);
                    });

                    if (variables && resolved) {
                        it(`should be able to resolve variables ${toString(variables)}`, () => {
                            const parser = new Parser(query, {
                                type_config: typeConfig,
                                filterNilVariables: true,
                                variables
                            }).resolveVariables(variables);

                            expect(parser.ast).toMatchObject(resolved);
                        });
                    }
                });
            });
        }
    });

    describe('when testing edge cases', () => {
        describe('given a gigantic query', () => {
            it('should be able to parse it', () => {
                const partOne = times(300, (n) => `a:${n}`).join(' OR ');
                const partTwo = times(200, (n) => `b:${n}`).join(' OR ');
                const partThree = times(500, (n) => `c:${n}`).join(') OR (');
                const parser = new Parser(`(${partOne}) AND ${partTwo} OR (${partThree})`);
                expect(parser.ast).toMatchObject({
                    type: NodeType.LogicalGroup,
                });
            });
        });
    });

    describe('when given a invalid query "(ba"', () => {
        it('should throw an error', () => {
            const errMsg = [
                'Failure to parse xLucene query "(ba",',
                'caused by SyntaxError: Expected ")", field, term, or whitespace but end of input found.',
            ].join(' ');

            expect(() => {
                new Parser('(ba');
            }).toThrowWithMessage(TSError, errMsg);
        });
    });

    describe('when given a invalid query "AND foo:bar"', () => {
        it('should throw an error', () => {
            const errMsg = 'Failure to parse xLucene query "AND foo:bar", caused by SyntaxError: Expected term, or whitespace but ":" found.';
            expect(() => {
                new Parser('AND foo:bar');
            }).toThrowWithMessage(TSError, errMsg);
        });
    });

    describe('when given a invalid query "OR foo:bar"', () => {
        it('should throw an error', () => {
            const errMsg = 'Failure to parse xLucene query "OR foo:bar", caused by SyntaxError: Expected term, or whitespace but ":" found.';
            expect(() => {
                new Parser('OR foo:bar');
            }).toThrowWithMessage(TSError, errMsg);
        });
    });

    describe('when given a invalid scoped query', () => {
        it('should throw if there are multiple dots', () => {
            expect(() => {
                new Parser('foo:@bar..baz');
            }).toThrow();
        });

        it('should throw an error if it is combined with &', () => {
            expect(() => {
                new Parser('foo:$@bar');
            }).toThrow();
        });

        it('should throw an error if it is combined with & and @ is escaped', () => {
            expect(() => {
                new Parser('field:$\\@example');
            }).toThrow();
        });
    });

    it('should be to iterate over all of the values', () => {
        const parser = new Parser([
            'a:$foo',
            'b:>=20',
            'a:($foo @bar)',
            'geo:geoBox(top_left:" 33.906320, -112.758421", bottom_right:$bottom_right)',
        ].join(' AND '), {
            type_config: {
                a: xLuceneFieldType.Integer,
                b: xLuceneFieldType.Integer,
                geo: xLuceneFieldType.GeoPoint,
            },
        });

        const nodes: [FieldValue<any>, TermLikeNode][] = [];
        parser.forEachFieldValue((value, node) => nodes.push([value, node]));
        expect(nodes).toMatchSnapshot();
    });

    test.each([
        { some: 'data' },
        Buffer.from('1234'),
        new Map(),
        new Set(),
        new Error(),
        [1, { other: 'stuff' }, 3]
    ])('should throw if given invalid variable value %p', (value) => {
        const query = 'foo: $bar';
        const typeConfig = { foo: xLuceneFieldType.String };

        const variables = { bar: value };

        expect(() => {
            new Parser(query, {
                type_config: typeConfig,
            }).resolveVariables(variables);
        }).toThrow();
    });
});
