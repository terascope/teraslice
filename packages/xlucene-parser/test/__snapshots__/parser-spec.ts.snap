// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser should be to iterate over all of the values 1`] = `
Array [
  Array [
    Object {
      "scoped": false,
      "type": "variable",
      "value": "foo",
    },
    Object {
      "field": "a",
      "field_type": "integer",
      "type": "term",
      "value": Object {
        "scoped": false,
        "type": "variable",
        "value": "foo",
      },
    },
  ],
  Array [
    Object {
      "type": "value",
      "value": 20,
    },
    Object {
      "field": "b",
      "left": Object {
        "field_type": "integer",
        "operator": "gte",
        "type": "term",
        "value": Object {
          "type": "value",
          "value": 20,
        },
      },
      "type": "range",
    },
  ],
  Array [
    Object {
      "scoped": false,
      "type": "variable",
      "value": "foo",
    },
    Object {
      "field": "a",
      "field_type": "integer",
      "type": "term",
      "value": Object {
        "scoped": false,
        "type": "variable",
        "value": "foo",
      },
    },
  ],
  Array [
    Object {
      "scoped": true,
      "type": "variable",
      "value": "@bar",
    },
    Object {
      "field": "a",
      "field_type": "integer",
      "type": "term",
      "value": Object {
        "scoped": true,
        "type": "variable",
        "value": "@bar",
      },
    },
  ],
  Array [
    Object {
      "type": "value",
      "value": " 33.906320, -112.758421",
    },
    Object {
      "field": "top_left",
      "field_type": "string",
      "quoted": true,
      "type": "term",
      "value": Object {
        "type": "value",
        "value": " 33.906320, -112.758421",
      },
    },
  ],
  Array [
    Object {
      "scoped": false,
      "type": "variable",
      "value": "bottom_right",
    },
    Object {
      "field": "bottom_right",
      "field_type": undefined,
      "type": "term",
      "value": Object {
        "scoped": false,
        "type": "variable",
        "value": "bottom_right",
      },
    },
  ],
]
`;
